src/components/cylc/Graph.vue

```diff
@@ -0,0 +1,1284 @@
+      }
+    },
+
+    // for websocket functionality
+    // async messageReceived (msg) {
+    //   console.log('graph view messageRecieved: ', msg)
+    //   this.graphData = JSON.parse(msg.data) // update via watcher
+    // },
+
+    changeLayout (value) {
+      this.layoutName = value
+    },
+
+    async handleMounted () {
+      try {
+        this.status = 'success'
+      } catch (error) {
+        console.error('handleMounted error: ', error)
+        this.status = 'error'
+      }
+    },
+
+    async preConfig (cytoscape) {
+      // cytoscape: this is the cytoscape constructor
+      try {
+        console.log('PRE-CONFIG')
+        cytoscape.use(cola)
+        cytoscape.use(dagre)
+        cytoscape.use(coseBilkent)
+        cytoscape.use(klay)
+        this.cy = cytoscape({
+          container: document.getElementById('cytoscape'),
+          elements: this.graphData
+        })
+        this.debouncer = debounce(this.updateGraph, 100)
+      } catch (error) {
+        console.error('preConfig error: ', error)
+      }
+    },
+
+    async afterCreated (cy) {
+      try {
+      // load graph data and run layout
+        layoutOptions = dagreOptions
+        expandCollapseOptions = expandCollapseOptionsUndefined
+        const loaded = await this.initialise(cy)
+        loaded ? this.loading = false : console.log('there was an error loading the graph view')
+      } catch (error) {
+        console.log('afterCreated error', error)
+      }
+    },
+
+    async registerExtensions (instance) {
+      try {
+        // register extensions
+        if (typeof cytoscape('core', 'navigator') !== 'function') {
+          console.log('registering navigator')
+          navigator(cytoscape)
+        }
+
+        if (typeof cytoscape('core', 'panzoom') !== 'function') {
+          console.log('registering panzoom')
+          panzoom(cytoscape)
+        }
+
+        if (typeof cytoscape('core', 'undoRedo') !== 'function') {
+          console.log('registering undoRedo')
+          undoRedo(cytoscape)
+        }
+
+        if (typeof cytoscape('core', 'expandCollapse') !== 'function') {
+          console.log('registering expandCollapse with jquery')
+          expandCollapse(cytoscape, jquery)
+        }
+
+        if (typeof cytoscape('core', 'popper') !== 'function') {
+          console.log('registering popper')
+          popper(cytoscape)
+        }
+      } catch (error) {
+        console.log('registerExtensions error', error)
+      }
+    },
+
+    async updateStyle (data) {
+      try {
+        return this.updateConfig(data)
+      } catch (error) {
+        console.log('updateStyle error: ', error)
+      }
+    },
+
+    async updateConfig (data) {
+      try {
+        const config = {
+          autounselectify: true,
+          boxSelectionEnabled: true,
+          layout: {
+            name: 'dagre',
+            textureOnViewport: false,
+            hideEdgesOnViewport: true
+          },
+          style: [
```

Comment:
Hmmm, looks like this is going to be tricky with Cytoscape. Spent some time today reading Cytoscape documentation, and playing with the code here. I believe now I understand what @MartinRyan said when he said we would have to re-compute the styles for each update.

## Not really CSS

There's some key information in their docs about styles https://js.cytoscape.org/#style:

>Style in Cytoscape.js follows CSS conventions as closely as possible. In most cases, a property has the same name and behaviour as its corresponding CSS namesake. However, the properties in CSS are not sufficient to specify the style of some parts of the graph. In that case, additional properties are introduced that are unique to Cytoscape.js.

So the style applied to nodes and edges is not necessarily CSS. Though it does look at lot like. This function that does part of the parsing of the style properties shows more or less what happens in the lower levels: https://github.com/cytoscape/cytoscape.js/blob/088995626b9cd69013da183980a3f178edc5a149/src/style/parse.js#L60

It has a list of properties that can be used... as in the docs :nerd_face: 

## But we can still define "classes"... though they must come during the initialisation

If we would like to have something like CSS classes, our best option is to use the Cytoscape [initialisation](https://js.cytoscape.org/#core/initialisation) and provide a `styles` to the Cytoscape object with the classes/groups that we want to have. e.g. (not tested)

```js
const styles = {
  '.running': {
    background-image: ...,
    background-color: states.RUNNING.color,
    ...
  },
  '.succeeded': {
    background-image: ...,
    background-color: states.SUCCEEDED.color,
    ...
  },
  ...
}
```

Doing this, in the `getNodes` and `getEdges`, we would then have to use the `classes: ['succeeded']`. It is similar to the current approach, but instead of providing functions that will receive the `node`, and then decide what to do, we specify the class and then it doesn't have to compute it again.

## cy.batch

We should probably try to use [`cy.batch`](http://js.cytoscape.org/#cy.batch) if we have performance issues due to recalculating styles.

## Theming

I think this affects theming. @oliver-sanders had had a look at some graph libraries before, and he has a very good understanding of how theming needs to work in Cylc, to allow us to support not only different nice UI's, but also attend users with accessibility issues (I know UK & NZ have users needing this).

So far, it seems to be that Cytoscape is not really compatible with our current approach for theming. Even though @oliver-sanders and I haven't agreed on how to implement it exactly, we both were looking at creating classes/variables in SCSS files, and allowing users to change it either in CSS or SCSS.

Looks like Cytoscape and SCSS won't work together in a component. So I wonder if @MartinRyan has an alternative for this, or if there's still time to investigate other libraries?

Reply:
These are definitely the big questions that we need to figure out. We do still need to investigate other libraries, particularly with consideration of our special requirements - that was always the intention, as we've noted a bunch of times.  I wonder why Cytoscape did styling this way?

Reply:
You could probably have different stylesheet options linked to the different themes, as with changing graph layouts. The less that is computed in the stylesheet the greater the performance.